File: java/revenj-servlet/src/main/java/org/revenj/server/CommandResult.java
package org.revenj.server;

public final class CommandResult<TFormat> {
	public final TFormat data;
	public final String message;
	public final int status;

	public CommandResult(TFormat data, String message, int status) {
		this.data = data;
		this.message = message;
		this.status = status;
	}

	public static <TFormat> CommandResult<TFormat> badRequest(String message) {
		return new CommandResult<>(null, message, 400);
	}

	public static <TFormat> CommandResult<TFormat> forbidden(String name) {
		return new CommandResult<>(null, "You don't have permissions to access:" + name, 403);
	}

	public static <TFormat> CommandResult<TFormat> success(String message, TFormat value) {
		return new CommandResult<>(value, message, 200);
	}
}

File: scala/revenj-akka/src/main/scala/net/revenj/server/CommandResult.scala
package net.revenj.server

import scala.concurrent.Future
import scala.util.Try

case class CommandResult[TFormat](data: Option[TFormat], message: String, status: Int)

object CommandResult {
  def badRequest[TFormat](message: String): Future[CommandResult[TFormat]] = {
    Future.successful(CommandResult[TFormat](None, message, 400))
  }

  def error[TFormat](message: String): Future[CommandResult[TFormat]] = {
    Future.successful(CommandResult[TFormat](None, message, 500))
  }

  def badRequest[TFormat](error: Try[Throwable]): Future[CommandResult[TFormat]] = {
    Future.successful(CommandResult[TFormat](None, error.get.getMessage, 400))
  }

  def forbidden[TFormat](name: String): Future[CommandResult[TFormat]] = {
    Future.successful(CommandResult[TFormat](None, s"You don't have permissions to access: $name", 403))
  }

  def success[TFormat](message: String, value: TFormat): Future[CommandResult[TFormat]] = {
    Future.successful(CommandResult(Some(value), message, 200))
  }
}
